### DISCUSSION TITLE
CodeQL Action won't build

### DISCUSSION THREAD

**Discussion by flagrant99** [community/community#66186] [Code Security] [answered]
### Select Topic Area

Question

### Body

I have a build workflow that works fine.
I created a codeQL work flow using the wizard but Autobuild fails. 
So I copy over my build code and replace the Autobuild and that build never succeeds. 

When comparing output I see these lines all over the place

EmitCompilerGeneratedFiles was true, but no CompilerGeneratedFilesOutputPath was provided. CompilerGeneratedFilesOutputPath must be set in order to emit generated files

Can someone fill me in on what CodeQL uses for static code analysis and what the "build" is actually supposed to create?
It ends up failing every time. 

All the build does is call a powershell file to do the build. There has to be something different in the CodeQL environment that is breaking things.

**Comment by aeisenberg** [comment]
Can you share your workflow file? It would be easier to understand what is happening if I can see it. (If this is from a public repository, then can you share the link to it?) Also, you can re-run the analysis in debug mode to capture more artifacts. If you attach the log, I can take a look at it and help you figure out what's going on.

CodeQL generates a database of your code by hooking into the compiler and reading compiler artifacts. We call this process "extraction". Autobuild will attempt to compile your code using some default, generic build commands that are specific for your language. Eg- for Java, it will try running `mvn compile` and if that fails, then `./gradlew run`, and then a few other commands. If your repo uses a non-standard build, then autobuild will fail. Instead, you need to supply your own build commands. CodeQL hooks into the compiler and listens for artifacts being generated. It then uses these artifacts to stuff things into the database.

(Side note: once the build is complete, CodeQL does not need to the build artifacts _or_ the source code any more since everything it needs is in the database.)

**Reply by flagrant99** [reply]
Searching D:\actions-runner\ on git build server for emitcompilergeneratedfiles I get some clues
D:\actions-runner\_work\_temp\codeql_databases\log\build-tracer.log

[T 09:55:46 9248] invocation: c:\program files\microsoft visual studio\2022\enterprise\msbuild\current\bin\msbuild.exe, args: 
/p:SolutionName="Integration" /p:UseSharedCompilation=false /p:MvcBuildViews=true /p:EmitCompilerGeneratedFiles=true

Now that I enabled MvcBuildViews in the csproj I commented out all the uncompilable web code we don't use. 

Now I get code scanning alerts for csharp. 

Is there away to see the list of files it scanned?

**Reply by aeisenberg** [reply]
That's great that you are able to get code scanning alerts. You should be able to see how many files were scanned vs how many source files are in your repository. You can see this at this link (replace ORG and REPO appropriately):

https://github.com/ORG/REPO/security/code-scanning/tools/CodeQL/status

If you are looking to find out exactly what files have been included in the database, you can download the database artifact from the build job and look at the `src.zip` file.

**Reply by flagrant99** [reply]
I see no artifacts from codeql.yml
So I added

    - uses: actions/upload-artifact@v3
      with:
        name: CodeQLArtifact
        path: ./build_output/


and that gives me
[Analyze (csharp)]
No files were found with the provided path: ./build_output/. No artifacts will be uploaded.

So what path should I give to get the database from and/or src.zip

**Reply by aeisenberg** [reply, marked as answer]
Sorry, I was mistaken. The analyze action will upload your database automatically after the analysis completes. But it will only do this on the default branch.

In your case, you will need to specify a `db-location` property in the `init` action and then upload the database as an artifact from there.

Something like this perhaps:

```yaml
- uses: github/codeql-action/init
  with:
    ...
    db-location: database
  ...
- uses: actions/upload-artifact@v3
  with:
    name: codeql-database
    path: database
  ...
```

**Comment by ShahbazShaddy** [comment]
It sounds like you're encountering issues with your CodeQL workflow and specifically with the build process. CodeQL is a powerful static code analysis tool, and it's commonly used to find security vulnerabilities and bugs in code.

The error message you're seeing, "EmitCompilerGeneratedFiles was true, but no CompilerGeneratedFilesOutputPath was provided," suggests that there might be a configuration issue with how the build process is set up in your CodeQL workflow.

Here are a few things you might consider:

1. **Check CodeQL Documentation**: Start by reviewing the official CodeQL documentation provided by GitHub. They often have detailed guides and troubleshooting tips.

2. **Verify Environment Variables**: Ensure that any environment variables that your build process relies on are correctly set up in your CodeQL workflow. This could include variables related to paths, dependencies, or build configurations.

3. **Check Build Logs**: Examine the build logs in detail. They may provide more specific information about where the process is failing and why. Look for any error messages or warnings that could shed light on the issue.

4. **Compare Environments**: Consider the differences between your local environment (where the build works fine) and the CodeQL environment. Are there any specific dependencies, configurations, or permissions that are required for the build process to succeed in CodeQL?

5. **Seek Community Help**: GitHub has a community forum and support channels where you can ask for help from other developers who might have encountered similar issues.

6. **Review Your PowerShell Script**: Double-check that your PowerShell script is compatible with the environment in which it's running. Ensure that it doesn't rely on any local resources or configurations that might not be available in the CodeQL environment.

7. **Reach Out to GitHub Support**: If all else fails, consider reaching out to GitHub support or their community forums. They may be able to provide more specific guidance based on the details of your workflow.

Without specific knowledge of your code, workflow, and environment, it's challenging to provide a precise solution. However, following these general steps should help you narrow down the issue and hopefully find a resolution.

Remember to make backups and version control your workflow files before making significant changes to avoid potential data loss.
